src/Workspaces/Core/Portable/Utilities/ComparerWithState.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Immutable;
+
+namespace Roslyn.Utilities
+{
+    internal delegate int ComparisonWithState<in T, in S>(T x, T y, S state);
+
+    /// <summary>
+    /// Extension of <see cref="Comparer"/> to take into account dynamic <typeparamref name="S"/>
+    /// </summary>
+    /// <typeparam name="T">Comparing type</typeparam>
+    /// <typeparam name="S">State</typeparam>
+    internal class ComparerWithState<T, S>
+    {
+        private readonly ComparisonWithState<T, S> _comparison;
+
+        internal ComparerWithState(ComparisonWithState<T, S> comparison)
+        {
+            _comparison = comparison;
+        }
+
+        public int Compare(T x, T y, S state)
+            => _comparison(x, y, state);
+
+        public static ComparerWithState<T, S> Create(Func<T, S, IComparable> comparingMethod)
+            => new ComparerWithState<T, S>((t1, t2, state) => comparingMethod(t1, state).CompareTo(comparingMethod(t2, state)));
+    }
+
+    internal class ComparerWithState<T> : ComparerWithState<T, object>
+    {
+        internal ComparerWithState(ComparisonWithState<T, object> comparison) : base(comparison)
+        {
+        }
+
+        public static ComparerWithState<T> Create(Func<T, IComparable> comparingMethod)
+            => new ComparerWithState<T>((t1, t2, state) => comparingMethod(t1).CompareTo(comparingMethod(t2)));
+    }
+
+    internal class ComparerWithState
```

Comment:
static class.

i can understand these helpers to build up a final comparer-thingy out of the individual pieces.  I don't quite get the need the for the above classes.